driver-core/src/main/java/com/datastax/driver/core/PercentileTracker.java

```diff
@@ -0,0 +1,339 @@
+ * <p/>
+ * This class is used by percentile-aware components such as
+ * {@link QueryLogger.Builder#withDynamicThreshold(PercentileTracker, double)}  QueryLogger} and
+ * {@link com.datastax.driver.core.policies.PercentileSpeculativeExecutionPolicy}.
+ * <p/>
+ * It uses <a href="http://hdrhistogram.github.io/HdrHistogram/">HdrHistogram</a> to record latencies:
+ * for each category, there is a "live" histogram where current latencies are recorded, and a "cached", read-only
+ * histogram that is used when clients call {@link #getLatencyAtPercentile(Host, Statement, Exception, double)}. Each
+ * time the cached histogram becomes older than the interval, the two histograms are switched. Statistics will not be
+ * available during the first interval at cluster startup, since we don't have a cached histogram yet.
+ */
+public abstract class PercentileTracker implements LifecycleAwareLatencyTracker {
+    private static final Logger logger = LoggerFactory.getLogger(PercentileTracker.class);
+
+    private final long highestTrackableLatencyMillis;
+    private final int numberOfSignificantValueDigits;
+    private final int minRecordedValues;
+    private final long intervalMs;
+
+    // The "live" recorders: this is where we store the latencies received from the cluster
+    private final ConcurrentMap<Object, Recorder> recorders;
+    // The cached histograms, corresponding to the previous interval. This is where we get the percentiles from when the
+    // user requests them. Each histogram is valid for a given duration, when it gets stale we request a new one from
+    // the corresponding recorder.
+    private final ConcurrentMap<Object, CachedHistogram> cachedHistograms;
+
+    /**
+     * Builds a new instance.
+     *
+     * @see Builder
+     */
+    protected PercentileTracker(long highestTrackableLatencyMillis,
+                                int numberOfSignificantValueDigits,
+                                int minRecordedValues,
+                                long intervalMs) {
+        this.highestTrackableLatencyMillis = highestTrackableLatencyMillis;
+        this.numberOfSignificantValueDigits = numberOfSignificantValueDigits;
+        this.minRecordedValues = minRecordedValues;
+        this.intervalMs = intervalMs;
+        this.recorders = new ConcurrentHashMap<Object, Recorder>();
+        this.cachedHistograms = new ConcurrentHashMap<Object, CachedHistogram>();
+    }
+
+    /**
+     * Computes a key used to categorize measurements. Measurements with the same key will be recorded in the same
+     * histogram.
+     *
+     * @param host      the host that was queried.
+     * @param statement the statement that was executed.
+     * @param exception if the query failed, the corresponding exception.
+     * @return the key, or {@code null} if we're not yet ready to record.
+     */
+    protected abstract Object computeKey(Host host, Statement statement, Exception exception);
```

Comment:
Since we maintain a Histogram & Recorder for each key, should we make note too be careful not to have too many possibly unique keys?